Optimize breadcrumb computation performances

jocel1 opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Greatly improve performances with a lot of categories
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-3666
| How to test?  | load a product page in BO with a lot of categories in the database

It actually improves both frontoffice & backoffice performances when a lot of categories are used

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8896)
<!-- Reviewable:end -->

jocel1 committed:
speedup category creations during install
speedup get parent category by using two queries overall instead of one query by parent.

jocel1 committed:
compute the breadcrumb infos in php by getting all the relevent categories info with only a 3 sql queries

jocel1 committed:
Coding style

jocel1 commented:
With about 2000 categories : 
https://blackfire.io/profiles/compare/aa083b56-1d01-47c9-b986-486684eb275c/graph

mickaelandrieu commented on classes/Category.php:
```diff
@@ -780,6 +787,45 @@ public static function getSimpleCategories($idLang)
+     *
+     * @param int $idLang Language ID
+     *
+     * @return array|false|mysqli_result|null|PDOStatement|resource
```
are you sure it is all types you may return? :trollface:

jocel1 replied:
mmm perhaps mysql_result is missing here :trollface:

mickaelandrieu commented on classes/Category.php:
```diff
@@ -1435,44 +1481,55 @@ public function getParentsCategories($idLang = null)
-        if (count(Category::getCategoriesWithoutParent()) > 1 && Configuration::get('PS_MULTISHOP_FEATURE_ACTIVE') && count(Shop::getShops(true, null, true)) != 1) {
+        if (count(Category::getCategoriesWithoutParent()) > 1
+            && Configuration::get('PS_MULTISHOP_FEATURE_ACTIVE')
+            && count(Shop::getShops(true, null, true)) != 1) {
```
```diff
- != 1
+ !== 1
```

mickaelandrieu commented on classes/Category.php:
```diff
@@ -1435,44 +1481,55 @@ public function getParentsCategories($idLang = null)
-                $sql .= ' AND c.`id_parent` != 0';
-            }
+                    AND `id_lang` = '.(int) $idLang.Shop::addSqlRestrictionOnLang('cl').')';
+        if (Shop::isFeatureActive() && Shop::getContext() == Shop::CONTEXT_SHOP) {
```
can you use "===" instead of "==" when possible?

jocel1 committed:
coding style: makes mickael happy

jocel1 commented:
With 12000 categories: https://blackfire.io/profiles/compare/4da5f0af-3bdd-4898-9233-4521842db8ed/graph
The number of sql queries is *a bit* lower now :trollface:

mickaelandrieu commented:
Hello @jocel1 can you explain how someone from QA team can test and validate your contribution?

Quetzacoalt91 commented on classes/Category.php:
```diff
@@ -1435,44 +1481,55 @@ public function getParentsCategories($idLang = null)
+        $rootTreeInfo = Db::getInstance(_PS_USE_SQL_SLAVE_)->getRow(
+            'SELECT c.`nleft`, c.`nright` FROM `'._DB_PREFIX_.'category` c 
+            WHERE c.`id_category` = '.(int) $context->shop->id_category
+        );
```
It seems this request is not mandatory `if (!empty($treeInfo))`, and can be moved in the `if` block?

Quetzacoalt91 commented on src/Adapter/Category/CategoryDataProvider.php:
```diff
@@ -167,23 +171,37 @@ public function getCategoriesWithBreadCrumb()
+    {
+        $categories = [];
+
+        while (isset($this->categoryList[$categoryId])) {
```
I wonder how you get out of the `while` statement? I don't see any `unset` or `= null` in this method.

jocel1 replied:
the $categoryId is updated at each loop to get the parent id value :)

jocel1 committed:
move the rootTreeInfo query at the right place

marionf commented:
@jocel1 

I have a lot of errors at the installation, maybe it's my local configuration

![capture du 2018-04-05 08-52-39](https://user-images.githubusercontent.com/13449658/38356761-e91a88a4-38c0-11e8-9c6d-058ba40c3da1.png)

jocel1 committed:
flush entities in the DB every 100 queries

alegout commented:
Reviewed 2 of 3 files at r1, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[classes/Category.php, line 91 at r4](https://beta.reviewable.io/reviews/prestashop/prestashop/8896#-L9odEEA2b0sLUUE9B7o:-L9odEEA2b0sLUUE9B7p:b-v7o6u1) ([raw file](https://github.com/prestashop/prestashop/blob/c62800e5a649c1d23389bae6d94cf1b61f71e61f/classes/Category.php#L91)):*
> ```PHP
> 
>     /** @var bool */
>     public $doNotRegenerateNTree = false;
> ```

Does it need to be public?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8896)*
<!-- Sent from Reviewable.io -->

jocel1 commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[classes/Category.php, line 91 at r4](https://beta.reviewable.io/reviews/prestashop/prestashop/8896#-L9odEEA2b0sLUUE9B7o:-L9oeV1G4Gu_HBfYWwfw:bjwyhbj) ([raw file](https://github.com/prestashop/prestashop/blob/c62800e5a649c1d23389bae6d94cf1b61f71e61f/classes/Category.php#L91)):*
<details><summary><i>Previously, alegout (Alban) wrote…</i></summary><blockquote>

Does it need to be public?
</blockquote></details>

yes, it was not defined at all before, and was hence public

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8896)*
<!-- Sent from Reviewable.io -->

alegout commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8896#-:-L9of7ufD-3PsfHHA88X:bnfp4nl)*
<!-- Sent from Reviewable.io -->

alegout reviewed (dismissed):
Thank you @jocel1 !

xmovu commented:
Thank you for the work, I appreciate, as I have over 14k categories and am unable to load products at all..

Now, I've merged this to my PrestaShop install, though it just throws me error 500s loading both categories and products, I'm unable to get any logs to the error_log as well. Any idea?

Edit, error editing the product (was only able to load product list in developer mode no idea why), attaches as productedit.txt
[productload.txt](https://github.com/PrestaShop/PrestaShop/files/1899377/productload.txt)

I can now load categories, but I cannot add or edit both categories and products.

Another issue editing products.
[editingproducts.txt](https://github.com/PrestaShop/PrestaShop/files/1899605/editingproducts.txt)


---

Renaming to the old CategoryDataProvider.php allows me to edit products. Although, I've cleared my store, and am reimporting stuff again, and for now, it works, but we'll see once it allows me to fully import categories and products. I was also getting an undisclosed error while importing categories.

mickaelandrieu commented:
@wdimd you may need to clear the cache?

xmovu commented:
I've done that many times, even entirely disabled it @mickaelandrieu . This got me the install completely messed up and will need to start from scratch.

Edit: Reimporting categories and products again, then will use the new files to see if it helps. I have been completely unable to add or edit both products and categories. I don't think this is an issue with the server, its load is barely 1, running php-fpm stack, on 8 core, 32GB RAM, RAID 10 NVMe.

Though I see it PrestaShop is having issues with 14k categories, and 7.4k products. Investigating, hopefully this fix will help me. Until now, it got nothing but issues but I could be doing something wrong.

xmovu commented:
Reinstalled it from scratch, imported 14k categories, didn't import the products yet, and I'm unable to create any products. Running the last PrestaShop from 9th April:

[11-Apr-2018 19:24:16 Europe/Lisbon] PHP Fatal error:  Out of memory (allocated 548220928) (tried to allocate 20480 bytes) in /home/dir/public_html/vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Core/Type/FormType.php on line 147

--
I've set no limit for memory usage, though this amount seems to be very exagerated, no?

Edit: So yeah, I've imported both categories and products now. Over 14k categories, roughly 7440 products. I am able to edit and add categories. But I am unable to add or edit products, it throws the memory error above, even though I did set the limit to -1, which is virtually no limit set. It's the first time we're getting Prestashop to production, as ink resellers, and this is damn crazy. Love prestashop and bought many plugins already, but I can't get it why it's unable to handle these products/categories. Maybe something in my end? I run php 7.1 btw.

mickaelandrieu commented:
ping @alegout wdyt? do you reproduce the bug?

jocel1 committed:
disable cache for the query which returns the categories as it could returns a lot of rows

jocel1 commented:
@wdimd I've just pushed a fix to avoid in your case to store the result of the new big category query into the prestashop query cache.
As for the memory usage, try to raise it, with such a high number of categories, it will have to create a lot of forms in the product page

xmovu commented:
@jocel1  I'll download the new code.
BTW, attached new information. Do note that I've set no limits for memory (it's -1), server has 32GB RAM. I even tried with 1024M, and it won't work to edit/add products.

Logs attached, I'll be trying your new code in  a few.
[prodlogs.txt](https://github.com/PrestaShop/PrestaShop/files/1900528/prodlogs.txt)

Edit, @jocel1 really weird stuff. I didn't notice any different, though, issue remains adding or editing a product. There must be a value, somewhere, dropping the memory limit, or defining it. I've loaded a php info to debug the ram and I've assigned 2048M, it tries to use: 
[2018-04-11 22:09:10] request.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\OutOfMemoryException: "Error: Out of memory (allocated 548220928) (tried to allocate 20480 bytes)

Doesn't make sense. I've 2GB assigned.

xmovu commented:
@jocel1,

Update: I'm able to add a product, or edit an existing one. Issue was Apache was limited to 1024M, and it wasn't enough, I've bumped it to 4096M, which is a lot.
Is still takes forever.
I'd say this is a damn good challenge, getting Prestashop to run properly when you have 7k+ products and 14k+ categories, where the products are assigned to multiple categories. There must be a way to get this much faster than it is. Takes forever to edit/add product, at least 1 minute for the webpage to load itself. This is on a NVMe server with plenty of IOPs...

I just tested editing a product, or adding, if I load the categories to assign to (and they're split properly), it just blocks completely.

jocel1 commented:
Avoid using -1, I saw several bugs, especially inside modules which was not correctly handling this value.
It's weird it takes forever in your case (because I have tested with 12k categories and a great amount of products), could you try to disable the prestashop cache ?

xmovu commented:
@jocel1 
-1 has been disabled long time ago as soon as I seen the issue with Apache, I've set a fair limit, 4GB for PHP and 4GB for Apache, it takes forever to load. It won't let me do anything in the product edit/add, even though it stops loading, like if it was loaded, but it's not, as soon as I click in Prices, or expand categories, it won't work, browser will block, basically.
I've just disabled cache, it won't allow me to do it anyway, changing to quantity / price tab blocks it entirely.

Edit: Measured with Chrome dev tools, 45s to load, 16MB transferred, 76 requests. Seems too much? I got only a few plugins installed, the frontend is superfast, backend too, the issue seems to rely in the products. I'm more than willing to share the environment as a test, should it be, and I appreciate the super help. We've spent several months organizing both the categories / products database to import to Prestashop, and now this is getting us completely stuck. It's not that categories have 1000 thousand in a single category, they're all split in sub categories, you can preview the categories at https://www.goprint.pt/content/6-tinteiros-e-acessorios , the plugin in the top loads those, using ajax, it's super fast. I can't get it why it's so slow in the backend adding products.

jocel1 commented:
are you in dev mode or in prod mode ? in defines.inc.php you should set 
define('_PS_MODE_DEV_', false); 
otherwise you'll get crap performances as well

xmovu commented:
@jocel1 , switched to dev mode, took 2 minutes to load adding a product, and still won't allow me to do anything once it loads.

I noticed a couple 404s as well, in the folder admin/_wdt. No dice, I'm hopeless.

jocel1 commented:
which Prestashop version you are using BTW, the develop branch?

xmovu commented:
@jocel1 I'm using 1.7.3.1, released in 9th April, with your modifications.

jocel1 commented:
could you give it a try directly with the develop branch? (and this patch of course)

xmovu commented:
@jocel1 will need to investigate if I can simply replace the files, never done it that way for Prestashop, though this was released 2 days ago, not sure if that'll make any difference.

jocel1 commented:
there are a few other performances fixes in the upcoming 1.7.4.0 (that's why I suggest to test the develop branch) which should be interesting to test in your case, as well as https://github.com/PrestaShop/PrestaShop/pull/8904. Anyway, your use case is interesting, you can contact me directly at jocelyn.fournier@softizy.com by mail or hangout, perhaps I could take a look directly at your datas to verify whether it's a new performance issue.

xmovu commented:
@jocel1 I believe I've already emailed you! :)

Will email right now.

Yep indeed, in my case, it seems weird why it won't work properly, it's not like the database is huge, it's not, at all.

xmovu commented:
FYI, got Prestashop there, dev branch won't even import categories with this fix. I'll email you the info.
Accessing catalog, gives me an error as well.

jocel1 committed:
- fix category import when force Id is not used
- make sure we only regenerate the NTree / clear the smarty cache at the end of the import
- disable the garbage collector on the product detail page because there could be a lot of objects on them and keeping the GC is counter-productive

jocel1 commented:
impact of the gc_disable() on the product detail page: 
https://blackfire.io/profiles/compare/4ee56e6f-9105-419b-acf1-c97e0b537a6a/graph

jocel1 committed:
safety check

xmovu commented:
I would like to thank @jocel1 for all the effort here, definitely, a contribution that was very much needed for PrestaShop. Will keep following up pulls that relate to high number of categories/products as it seems there's plenty of room for improvement here.

CallejaMotos commented:
hi guys,
How can I change my files to make work my website with this problem?
Even if you know a fast fix as not charging categories when trying to access to product detail in backoffice could be ok for me. 
I follow all your steps in commit but I have more errors now.
Please, help.
